### DISCUSSION TITLE
Rule for Confirming Header Exists in OpenAPI Spec

### DISCUSSION THREAD

**Discussion by blueksy1012** [stoplightio/spectral#2291] [Q&A] [answered]
Hello everyone. I am trying to write a custom rule in Spectral to ensure that requests use a "Ocp-Apim-Subscription-Key" parameter in the header. I am having trouble getting the rule to work. Can someone please help me?

This is what my custom rule looks like:
`  requests-include-authorization:
    description: "Requests must specify subsciprition key in headers"
    message: "Requests must specify subsciprition key in headers"
    severity: error
    given: $..parameters.[?(@.in === 'header')].name
    then:
        field: "Ocp-Apim-Subscription-Key"
        function: truthy`

I have attached part of
[ApiSpec.txt](https://github.com/stoplightio/spectral/files/9620171/ApiSpec.txt)
 the open api spec (.yaml) file to the discussion. 

I want my custom rule to apply when I remove the "Ocp-Apim-Subscription-Key" parameter from line 20. But it's not working.
My custom rule file is also attached here: 
[customrules.txt](https://github.com/stoplightio/spectral/files/9620194/customrules.txt)

Any help would be greatly appreciated. Thanks.

**Comment by EPKLISN** [comment, marked as answer]
I like to do it this way:

### the rule ###
given:
      - $.paths..parameters
    severity: hint
    then:
      function: parameters-header
      functionOptions:
        name: Content-Location
        type: header


### the function ###
module.exports = (targetVal, opts) => { 
  const { name } = opts;
  const { type } = opts;
  if (targetVal) {
    for (let i = 0; i < targetVal.length; i++) {
      if (targetVal[i].in == type && targetVal[i].name == name) {
        return;
      }
    }     
  return [{ message: 'Missing header parameter.' }];
  }
}

**Reply by blueksy1012** [reply]
Thank you for the response! I had wound up doing something similar with a custom function: 

`requests-include-authorization:
    message: "Requests {{error}}"
    severity: error
    given: $.paths..parameters
    then:
        function: has-header
        functionOptions:
          name: Ocp-Apim-Subscription-Key`

`export default  (input, opts) =>  {
  var hasHeader = false

  for (let i = 0; i < input.length; i++) {
    if(input[i].name == opts.name){
      hasHeader = true;
    }
  }

  if(!hasHeader){
      return [
          {
              message: "must include " + opts.name + " in the header",
          },
      ];
  }
};`
